### PR TITLE
feature add chart template and value api

### PR DESCRIPTION
**Pull request by zzhzero**

### CODE REVIEW

File: pkg/chartmuseum/server/multitenant/handlers.go [hunk 1]
```diff
@@ -116,7 +117,57 @@ func (server *MultiTenantServer) getStorageObjectRequestHandler(c *gin.Context)
...
+	c.JSON(200, map[string]interface{}{
+		"templates": chrt.Templates,
+		"values":    chrt.Values,
+	})
+}
+func (server *MultiTenantServer) getStorageObjectValuesRequestHandler(c *gin.Context) {
+	repo := c.Param("repo")
+	name := c.Param("name")
+	version := c.Param("version")
+	filename := fmt.Sprintf("%s-%s.tgz", name, version)
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/handlers.go:146]: there's an [existing function](https://github.com/helm/chartmuseum/blob/main/pkg/repo/chart.go#L45) for getting the chart filename that would be good to reuse here

**Comment by scbizu** [comment]
The PR relates to #509 ?

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 2]
```diff
@@ -1006,6 +1006,32 @@ func (suite *MultiTenantServerTestSuite) testAllRoutes(repo string, depth int) {
...
 
+	// GET /api/:repo/charts/:name/:version/templates
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/mychart/0.1.0/templates", apiPrefix), nil, "")
+	suite.Equal(200, res.Status(), fmt.Sprintf("200 GET %s/charts/mychart/0.1.0/templates", apiPrefix))
+
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/mychart/latest/templates", apiPrefix), nil, "")
+	suite.Equal(200, res.Status(), fmt.Sprintf("200 GET %s/charts/mychart/latest/templates", apiPrefix))
+
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/mychart/0.1.1/templates", apiPrefix), nil, "")
+	suite.Equal(404, res.Status(), fmt.Sprintf("200 GET %s/charts/mychart/0.1.1/templates", apiPrefix))
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/server_test.go:1017]:
```suggestion
	suite.Equal(404, res.Status(), fmt.Sprintf("404 GET %s/charts/mychart/0.1.1/templates", apiPrefix))
```

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 3]
```diff
@@ -1006,6 +1006,32 @@ func (suite *MultiTenantServerTestSuite) testAllRoutes(repo string, depth int) {
...
+
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/fakechart/0.1.0/templates", apiPrefix), nil, "")
+	suite.Equal(404, res.Status(), fmt.Sprintf("200 GET %s/charts/fakechart/0.1.0/templates", apiPrefix))
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/server_test.go:1020]:
```suggestion
	suite.Equal(404, res.Status(), fmt.Sprintf("404 GET %s/charts/fakechart/0.1.0/templates", apiPrefix))
```

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 4]
```diff
@@ -1006,6 +1006,32 @@ func (suite *MultiTenantServerTestSuite) testAllRoutes(repo string, depth int) {
...
+
+	// GET /api/:repo/charts/:name/:version/values
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/mychart/0.1.0/values", apiPrefix), nil, "")
+	suite.Equal(200, res.Status(), fmt.Sprintf("200 GET %s/charts/mychart/0.1.0/values", apiPrefix))
+
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/mychart/latest/values", apiPrefix), nil, "")
+	suite.Equal(200, res.Status(), fmt.Sprintf("200 GET %s/charts/mychart/latest/values", apiPrefix))
+
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/mychart/0.1.1/values", apiPrefix), nil, "")
+	suite.Equal(404, res.Status(), fmt.Sprintf("200 GET %s/charts/mychart/0.1.1/values", apiPrefix))
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/server_test.go:1030]:
```suggestion
	suite.Equal(404, res.Status(), fmt.Sprintf("404 GET %s/charts/mychart/0.1.1/values", apiPrefix))
```

File: pkg/chartmuseum/server/multitenant/server_test.go [hunk 5]
```diff
@@ -1006,6 +1006,32 @@ func (suite *MultiTenantServerTestSuite) testAllRoutes(repo string, depth int) {
...
+
+	res = suite.doRequest(stype, "GET", fmt.Sprintf("%s/charts/fakechart/0.1.0/values", apiPrefix), nil, "")
+	suite.Equal(404, res.Status(), fmt.Sprintf("200 GET %s/charts/fakechart/0.1.0/values", apiPrefix))
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/server_test.go:1033]:
```suggestion
	suite.Equal(404, res.Status(), fmt.Sprintf("404 GET %s/charts/fakechart/0.1.0/values", apiPrefix))
```

File: pkg/chartmuseum/server/multitenant/api.go [hunk 6]
```diff
@@ -96,7 +96,20 @@ func (server *MultiTenantServer) deleteChartVersion(log cm_logger.LoggingFn, rep
 	server.StorageBackend.DeleteObject(provFilename) // ignore error here, may be no prov file
 	return nil
 }
-
+func (server *MultiTenantServer) getChartFileName(log cm_logger.LoggingFn, repo string, name string, version string) (string, *HTTPError) {
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/api.go:99]: small nit: whitespace around this function

File: pkg/chartmuseum/server/multitenant/api.go [hunk 7]
```diff
@@ -96,7 +96,20 @@ func (server *MultiTenantServer) deleteChartVersion(log cm_logger.LoggingFn, rep
...
+	chartVersion, err := server.getChartVersion(log, repo, name, version)
+	if err != nil {
+		return "", err
+	}
+	if len(chartVersion.URLs) == 0 {
+		return "", &HTTPError{http.StatusNotFound, "not found chart filename"}
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/api.go:105]:
```suggestion
		return "", &HTTPError{http.StatusNotFound, "chart filename not found"}
```

File: pkg/chartmuseum/server/multitenant/api.go [hunk 8]
```diff
@@ -96,7 +96,20 @@ func (server *MultiTenantServer) deleteChartVersion(log cm_logger.LoggingFn, rep
...
+	}
+	split := strings.Split(chartVersion.URLs[0], "/")
+	if len(split) < 2 {
+		return "", &HTTPError{http.StatusNotFound, "not found chart filename"}
```
Comment (cbuto) [inline pkg/chartmuseum/server/multitenant/api.go:109]:
```suggestion
		return "", &HTTPError{http.StatusNotFound, "chart filename not found"}
```

**Review by cbuto — COMMENTED** [review]
few small nits around style

File: pkg/chartmuseum/server/multitenant/handlers.go [hunk 9]
```diff
@@ -19,6 +19,7 @@ package multitenant
 import (
 	"bytes"
 	"fmt"
+	"helm.sh/helm/v3/pkg/chart/loader"
```
Comment (scbizu) [inline pkg/chartmuseum/server/multitenant/handlers.go:22]: package imported here should be fmted .

File: pkg/chartmuseum/server/multitenant/handlers.go [hunk 10]
```diff
@@ -116,7 +117,70 @@ func (server *MultiTenantServer) getStorageObjectRequestHandler(c *gin.Context)
...
+	}
+	chrt, err1 := loader.LoadArchive(bytes.NewReader(storageObject.Content))
+	if err1 != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err1})
+		return
+	}
+	var data []byte
+	for _, file := range chrt.Raw {
+		if file.Name == "values.yaml" {
+			data = file.Data
```
Comment (scbizu) [inline pkg/chartmuseum/server/multitenant/handlers.go:175]: Since we only need to seek the values , it seems that we can break here maybe ?

**Review by scbizu — CHANGES_REQUESTED** [review]

**Comment by scbizu** [comment]
Another question , should this new handler be protected with the auth mechanism? 

/cc @cbuto

**Comment by cbuto** [comment]
@scbizu we definitely need auth for these endpoints if they aren't already covered 👀

**Review by cbuto — APPROVED** [review]
lgtm after @scbizu's concerns are addressed! 🎉

**Comment by cbuto** [comment]
want to give it another look @scbizu?

**Review by scbizu — APPROVED** [review]

**Comment by scbizu** [comment]
LGTM  , let me do the re-format things , thank you @zzhzero 🎉

**Pull request merged**